src/connectors/infinite-hits/connectInfiniteHits.ts

```diff
@@ -78,26 +138,30 @@ const connectInfiniteHits: InfiniteHitsConnector = function connectInfiniteHits(
       escapeHTML = true,
       transformItems = (items: any[]) => items,
       showPrevious: hasShowPrevious = false,
+      cache = getInMemoryCache(),
     } = widgetParams || ({} as typeof widgetParams);
-    let hitsCache: Hits = [];
-    let firstReceivedPage = Infinity;
-    let lastReceivedPage = -1;
+    let cachedHits: InfiniteHitsCachedHits | undefined = undefined;
     let prevState: Partial<SearchParameters>;
     let showPrevious: () => void;
     let showMore: () => void;
 
+    const getFirstReceivedPage = () =>
+      Math.min(...Object.keys(cachedHits || {}).map(Number));
+    const getLastReceivedPage = () =>
+      Math.max(...Object.keys(cachedHits || {}).map(Number));
+
     const getShowPrevious = (helper: Helper): (() => void) => () => {
       // Using the helper's `overrideStateWithoutTriggeringChangeEvent` method
       // avoid updating the browser URL when the user displays the previous page.
       helper
         .overrideStateWithoutTriggeringChangeEvent({
           ...helper.state,
-          page: firstReceivedPage - 1,
+          page: getFirstReceivedPage() - 1,
         })
         .search();
     };
     const getShowMore = (helper: Helper): (() => void) => () => {
-      helper.setPage(lastReceivedPage + 1).search();
+      helper.setPage(getLastReceivedPage() + 1).search();
```

Comment:
we are calculating this multiple times now, is that expected?

Reply:
When a new page is loaded, the `cachedHits` object will be filled with the hits.

```js
        if (cachedHits![page] === undefined) {
          cachedHits![page] = results.hits;
          cache.write({ state, hits: cachedHits });
        }
```

So `getShowMore` function needs to calculate it every time.